### PR TITLE
packfile: delta diff implementation

### PR DESCRIPTION
**Pull request by ajnavarro**
- Renamed delta.go to patch_delta.go
- Added diff_delta.go file
- Added tests that creates a diff and then tries to patch it

### CODE REVIEW

File: plumbing/format/packfile/diff.go [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+// Code obtained from https://github.com/pmezard/go-difflib
```
Comment (smola) [inline plumbing/format/packfile/diff.go]: Can't we import pmezard/go-difflib and use it directly?

Comment (smola) [inline plumbing/format/packfile/diff.go]: "Code obtained from" -> "Code based on"

File: plumbing/format/packfile/diff.go [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+	IsJunk         func(byte) bool
+	autoJunk       bool
+	bJunk          map[byte]struct{}
+	matchingBlocks []Match
+	fullBCount     map[byte]int
+	bPopular       map[byte]struct{}
+	opCodes        []OpCode
+}
+
+func NewMatcher(a, b []byte) *sequenceMatcher {
```
Comment (smola) [inline plumbing/format/packfile/diff.go]: private

File: plumbing/format/packfile/diff.go [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+// Code obtained from https://github.com/pmezard/go-difflib
+
+type Match struct {
```
Comment (smola) [inline plumbing/format/packfile/diff.go]: private

File: plumbing/format/packfile/diff.go [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+	A    int
+	B    int
+	Size int
+}
+
+type OpCode struct {
```
Comment (smola) [inline plumbing/format/packfile/diff.go]: private

File: plumbing/format/packfile/diff.go (hunk 1 above)
Comment (smola) [inline plumbing/format/packfile/diff.go]: Explain changes.

File: plumbing/format/packfile/diff_delta.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+// See https://github.com/jelmer/dulwich/blob/master/dulwich/pack.py and
+// https://github.com/tarruda/node-git-core/blob/master/src/js/delta.js
+// for more info
+
+const (
+	maxCopyLen = 0xffff
+)
+
+// GetDelta returns the way of how to transform base object to target object
+func GetDelta(base, target plumbing.Object) ([]byte, error) {
```
Comment (alcortesm) [inline plumbing/format/packfile/diff_delta.go]: This should receive `io.Reader`s, not `plubing.Object`s.

Comment (alcortesm) [inline plumbing/format/packfile/diff_delta.go]: Not sure we are following the correct approach here, by including modified files from other projects, let's discuss this.

Comment (smola) [inline plumbing/format/packfile/diff_delta.go]: Yes, please, use io.Reader.

Comment (smola) [inline plumbing/format/packfile/diff_delta.go]: Ok. GetDelta takes plumbing.Object in order to be symmetric with ApplyDelta. Let's keep it this way.

Comment (ajnavarro) [inline plumbing/format/packfile/diff_delta.go]: This method is to have the same logic that was used into patch_delta.go:

```
func ApplyDelta(target, base plumbing.Object, delta []byte) error {
```

**Pull request merged**